Improve typing performance by splitting attributes in state tree.

sgomes opened the pull request:
## Description
Attributes have been moved to `attributesByClientId` in order to reduce the impact of typing throughout the state tree (See #11782).

Together with this, several selectors have been rewritten to take advantage of this separation by depending only on `byClientId`, which is now a less expensive branch of the state tree.

## How has this been tested?
Unit tests and manual verification.

Performance has been tested by recording the quick typing of a short word in a large document and measuring the time taken handling each input event.

## Screenshots 
Not applicable.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [X] My code follows the accessibility standards.
- [X] My code has proper inline documentation.

sgomes committed:
Improve typing performance by splitting attributes in state tree.

Attributes have been moved to `attributesByClientId` in order to
reduce the impact of typing throughout the state tree.
See #11782.

Review fixes pass #1

Simplify block flattening code in reducers.

Remove alignment toolbar optimization; should be a different PR

Fix minor bug in test

Fix failing getBlockDependantsCacheBust test

Remove new `*WithoutAttributes` selectors.

We'll go with a different approach: use the existing selectors, but
keep the dependencies as they are. The attributes may get stale, but
it doesn't matter if they're not being used.

Change attributesById structure to not have an inner attributes

Simplifying some selector dependencies

Simplify withSaveReusableBlock a bit further and clarify its existence

Reuse constant instead of running omit again

Remove no longer needed mapClientIds

Simplifying reducers after review comments.

Further changes to selectors after review

Fix types in JSDoc

Renaming attributesByClientId to attributes

Further selector fixes after review

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,12 +436,74 @@ export const editor = flow( [
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							...omit( action.updates, 'attributes' ),
```
In my PR, I did a little optimization here where I avoid generating a new state if the only key in `action.updates` is `attributes`

sgomes replied:
Good idea! Added that in.

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,12 +436,74 @@ export const editor = flow( [
+
+				case 'UPDATE_BLOCK':
+					// Ignore updates if block isn't known
+					if ( ! state[ action.clientId ] ) {
```
We should avoid generating a new state here if there's no attributes updates in `action.updates`

sgomes replied:
Yup, another useful early return.

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,12 +436,74 @@ export const editor = flow( [
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							attributes: { ...action.updates.attributes },
```
The `attributes` key is redundant here, we can just remove it and update the state to be `{ [ clientID ]: attributes }` instead of `{ [ clientID ]: { attributes } }`

youknowriad replied:
Not a blocker for me.

aduth replied:
> I opted for maintaining the attributes key as a way of having the same general shape in both `byClientId` and `attributesByClientId`, but I'm not married to that choice.

But they're not meant to be the same sort of thing. The former is a block object, the latter (by its name) is exclusively attributes.

What else are we planning to include here? The spread of `...state[ action.clientId ]` seems to imply that there might be other non-`attribute` properties.

Given the name `attributesByClientId`, I would assume the contents of the object are in-fact attributes.

sgomes replied:
Done in latest commit.

youknowriad commented on packages/editor/src/store/selectors.js:
```diff
@@ -439,7 +440,7 @@ export function isEditedPostSaveable( state ) {
  */
 export function isEditedPostEmpty( state ) {
-	const blocks = getBlocksForSerialization( state );
+	const blocks = getBlocksForSerializationWithoutAttributes( state );
```
For more clarity, maybe this could be refactored to return just the blockNames : `getBlocksNamesForSerialization`

sgomes replied:
Changed. Also removed the export.

youknowriad commented on packages/editor/src/store/selectors.js:
```diff
@@ -659,7 +682,42 @@ export const getBlock = createSelector(
+ *
+ * @return {Object} Parsed block object.
+ */
+export const getBlockWithoutAttributes = createSelector(
```
I'm not certain why we need all these new selectors `getBlockWithoutAttributes` `getBlocksWithoutAttributes`, `getBlocksByClientIdWithoutAttributes`. I'd like to avoid exposing new selectors in this PR. Is this only for `getBlocksForSerializationWithoutAttributes`?

sgomes replied:
It gets used directly or indirectly in a bunch of other selectors (such as `getBlockNamesForSerialization`, `getInserterItems`, etc.) , but you're right that we don't need to export these new selectors.

sgomes replied:
The latest commit removes the new selectors, in exchange for the approach of using the same full selectors everywhere, but changing the dependencies where possible. This way, they won't invalidate on attribute changes, but that's ok if the attributes are not being used.

youknowriad commented on packages/editor/src/store/selectors.js:
```diff
@@ -1944,7 +2085,7 @@ export const getInserterItems = createSelector(
 	( state, rootClientId ) => [
 		state.blockListSettings[ rootClientId ],
-		state.editor.present.blocks,
+		state.editor.present.blocks.byClientId,
```
I think we probably need `.order` here as well right?

youknowriad replied:
Actually, I'm not certain

sgomes replied:
From what I can tell, block order should affect inserter items, but I could be wrong. Is there someone with more domain-specific knowledge of inserter items that could weigh in?

gziolo replied:
Nested blocks might have some impact on that, but as noted in the other PR, I would love us to explore whether we can skip it here.

sgomes replied:
I'm not too concerned with the performance implications of depending on `.order`, from what I've seen, so I'll add in that dependency just to be on the safe side. If this were to produce any issues due to the intricate dependencies you're describing, it could get extremely difficult to debug, so it's not worth the risk.

aduth replied:
> so I'll add in that dependency just to be on the safe side.

It still needs to be added?

sgomes replied:
Had forgotten to push my changes.

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -112,6 +112,33 @@ function getFlattenedBlocks( blocks ) {
+		[ blockId ]: { attributes: flattenedBlocks[ blockId ].attributes },
+	} ), {} );
+}
+
```
Great idea! Done.

sgomes commented on packages/editor/src/components/alignment-toolbar/index.js:
```diff
@@ -61,12 +62,10 @@ export function AlignmentToolbar( { isCollapsed, value, onChange, alignmentContr
 	);
 }
 
+const getClientIdOnly = memize( ( clientId ) => ( { clientId } ) );
```
I'll remove this optimization for now; it's distracting from the main purpose of the PR anyway.

sgomes commented:
All tests should now pass, and the code should be ready for a full review.

Addressed all existing comments and fixed everything except for removing the nesting of `attributes` inside `attributesByClientId`. @youknowriad: if you'd still like me to go ahead with that let me know, and I'll create a new commit in this PR for that.

sgomes commented:
After the fixes, particularly adding back cache busting, the performance improvement sits at ~17%.

youknowriad commented:
Unit tests are passing but the e2e tests are not. I'll see if I can find why.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,12 +432,80 @@ export const editor = flow( [
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							...omit( action.updates, 'attributes' ),
```
This value is already assigned above as `changes`. Avoiding a second `omit` would be ideal.

sgomes replied:
Good catch!

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,12 +432,80 @@ export const editor = flow( [
+			return state;
+		},
+
+		attributesByClientId( state = {}, action ) {
```
Bikeshedding: The naming of the properties of `blocks` already doesn't feel very consistent, and I'm worried this moves us further in that direction. If `order` is also keyed by client IDs, why not name it `orderByClientId` ? Or conversely, name this one just `attributes` ? What _are_ the contents `byClientId` anyways, particularly relevant to the proposed changes of pulling some values out. It makes me wonder if we'd be better served by granularity on all of the properties, rather than some ambiguous "byClientId" block-ish object shape, i.e. 

```graphql
{
	blocks: {
		isValid, # Object<string, boolean>
		name, # Object<string, string>
		attributes, # Object<string, Object>
		order, # Object<string, Array>
	}
}
```

youknowriad replied:
I think this is probably a good idea :) each property is mapped by ClientID, making it easier to memoize selectors based on the properties used. Maybe better as a follow-up to keep the size of this PR contained?

aduth replied:
> Maybe better as a follow-up to keep the size of this PR contained?

I raised because I think the proposed changes shine light on the fact that we might be better served by considering _all_ of these block values in the same way (hash keyed by client ID), which could lead us to some solution for generalizing the reducer implementation rather than copy/paste-ing large parts of it across as we're doing here. I'd worry we'd continue this pattern in some possible future refactor if set as a precedent.

sgomes replied:
I'm happy to make the change either in this PR or in a follow-up one. Let me know the final decision (when and which of the proposed conventions to go for) and I'll make it happen :)

youknowriad replied:
Let's try to do it here then.

sgomes replied:
`originalContent` seems to also be a common property for a block. Does that need to be extracted to another by-client-id list? Are there any others I should know about?

If we're bikeshedding, let's be thorough :)

sgomes replied:
As a more general comment, following the proposed approach would essentially encode the various properties of a block as separate branches in the state tree. The benefit is performance and separation, as discussed, but this comes at the cost of more difficulty in adding more properties to `block`, I would say.

As someone who is completely new to the project, I really don't know how to evaluate whether that's potentially a problem or not, but I wanted to bring it up as it occurred to me.

aduth replied:
> As a more general comment, following the proposed approach would essentially encode the various properties of a block as separate branches in the state tree. The benefit is performance and separation, as discussed, but this comes at the cost of more difficulty in adding more properties to `block` in the future (or changing them), I would say.

In my mind, I'd hoped it'd be some way we could programmatically generate the objects (reducers) based on some set of expected keys (`isValid`, `name`, `attributes`, and I suppose `originalContent` and any others), and that by doing it generically we would avoid any incurred cost in adding additional properties in the future, since it'd just be a key to add to the set.

This would assume that the logic for handling each key is consistent, which I'm not entirely confident is something we can assume (at least for attributes?). In which case, the refactoring may be more difficult and thus potentially not worthwhile. 

The big issue for me was in evaluating this part of state, how would you or anyone else answer the prompt "What is represented in the value of `byClientId`?" It's making less sense over time as we pull out separate properties to be able to call it as anything resembling a block object.

aduth replied:
More practically speaking, in my mind I'd had in mind some higher-order reducer akin to [`onSubKey`](https://github.com/WordPress/gutenberg/blob/80108fb9a2a5a2d3d5eb3692d0b47a65a9bb2073/packages/core-data/src/utils/on-sub-key.js), specific to handling the reducer actions relevant for blocks, on each block key.

So something roughly like:

```js
BLOCK_KEYS = [ 'name', 'isValid', 'originalContent', 'attributes' /* ? */ ]
blocks: combineReducers( BLOCK_KEYS.map( ( key ) => [ key, onBlockSubKey( key ) ] ) ) )
```

aduth replied:
> If `order` is also keyed by client IDs, why not name it `orderByClientId` ? Or conversely, name this one just `attributes` ?

Even if we're deferring a greater refactor, I'm still inclined to align these two. Thoughts on renaming to just `attributes`?

sgomes replied:
Sure, no problem! I have no strong opinions there. Added a commit with the change.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -754,6 +766,11 @@ export const getGlobalBlockCount = createSelector(
 
+const mapClientIds = ( clientIds, fn ) => map(
+	castArray( clientIds ),
+	( clientId ) => fn( clientId )
```
Minor: This line could be just `fn` ?

sgomes replied:
The method doesn't exist anymore.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -754,6 +766,11 @@ export const getGlobalBlockCount = createSelector(
 	]
 );
 
+const mapClientIds = ( clientIds, fn ) => map(
```
Not loving a randomly-placed, un-documented, un-tested utility function located here. Thinking either placement in a utilities file, or placement at top of function to have some arrangement of selectors vs. non-selectors. Is it just handling the coercion of `castArray`?

I don't see a ton of value in pulling this out as a function vs. just inlining.

sgomes replied:
Removed, no longer needed.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -1140,7 +1156,7 @@ const isAncestorOf = createSelector(
 	},
 	( state ) => [
-		state.editor.present.blocks,
+		state.editor.present.blocks.order,
```
Are these revisions relevant to the proposed change? Or are these things which are technically separate issues?

youknowriad replied:
Actually, these changes are one of the important changes in this PR as we don't invalidate these selectors unless the inserted blocks (order) change instead of any attribute change.

aduth replied:
I can see it would have an important impact on the overall benefit of the memoization; but is it specifically related to pulling out attributes to a separate state path?

youknowriad replied:
Actually, you're right, this one in particular is not directly related.

sgomes replied:
I'm happy enough to create a separate PR for this, but it doesn't seem like it would be outlandish to keep it in this PR either.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
 };
 
+/**
+ * Higher-order reducer which targets the combined blocks reducer and handles
```
What does "handles" entail? It's not immediately obvious to me as a reviewer now, let alone a future hypothetical maintainer.

youknowriad replied:
To clarify a bit: This HoR only updates the attributes of blocks not the "byClientId" part but we need to access "byClientId" to know the name of the blocks as we only want to change the `core/block` blocks.

sgomes replied:
I'll add a clarification that this HoR exists because it needs access to both branches of the state tree simultaneously.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+		newState.attributesByClientId = {};
+
+		Object.keys( state.byClientId ).forEach( ( blockId ) => {
+			const block = { ...state.byClientId[ blockId ] };
```
Why do we create shallow clones?

sgomes replied:
That's a good question. This must have stayed behind as I was figuring out the reducer. Sorry about that. Fixed!

sgomes commented on packages/editor/src/store/selectors.js:
```diff
@@ -691,9 +694,18 @@ export const getBlocks = createSelector(
-	( state ) => [
-		state.editor.present.blocks,
-	]
+	( state, rootClientId ) => {
```
Modified this to reuse the dependants of `getBlock` if a `rootClientId` is provided.

sgomes commented on packages/editor/src/store/selectors.js:
```diff
@@ -764,15 +781,12 @@ export const getGlobalBlockCount = createSelector(
 		state.initialEdits.meta,
 		state.currentPost.meta,
-		state.editor.present.blocks,
+		...mapClientIds( clientIds, ( clientId ) => getBlock( state, clientId ) ),
```
No problem. If this affects things too much we can always change it back.

noisysocks reviewed (commented):
This is working well in my testing. Nice work! 😍 

I left some mostly stylistic comments.

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+		newState.attributesByClientId = {};
+
+		Object.keys( state.byClientId ).forEach( ( blockId ) => {
+			const block = newState.byClientId[ blockId ];
```
Thanks! I don't think I had ever used lodash's `forEach` before.

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+			return state;
+		}
+
+		newState = { ...state };
```
I over-engineered this selector as I was figuring out what it was doing, and never went back to see if I could rewrite it in a simpler way. Thanks for that, it's much better! 👍

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+		newState = { ...state };
+		newState.attributesByClientId = {};
+
+		Object.keys( state.byClientId ).forEach( ( blockId ) => {
```
👍

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+
+		Object.keys( state.byClientId ).forEach( ( blockId ) => {
+			const block = newState.byClientId[ blockId ];
+			let attributes = { ...state.attributesByClientId[ blockId ] };
```
Fixed as part of using `mapValues`.

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -273,6 +300,48 @@ const withBlockReset = ( reducer ) => ( state, action ) => {
+ * @return {Function} Enhanced reducer function.
+ */
+const withSaveReusableBlock = ( reducer ) => ( state, action ) => {
+	let newState = state;
```
Yup, absolutely right. If we're just reassigning, we don't need the extra variable.

sgomes commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,63 +432,114 @@ export const editor = flow( [
-					if ( nextAttributes === state[ action.clientId ].attributes ) {
+					// Do nothing if only attributes change.
+					const changes = omit( action.updates, 'attributes' );
+					if ( keys( changes ).length === 0 ) {
```
Done.

sgomes commented on packages/editor/src/store/selectors.js:
```diff
@@ -1069,6 +1078,15 @@ export const getMultiSelectedBlockClientIds = createSelector(
 	],
 );
 
+const mapMultiSelectedBlockClientIds = ( state, fn ) => {
```
There's no advantage anymore. This was being used in multiple places, but after other changes, that's no longer the case. Good catch! Moving into `getMultiSelectedBlocks`.

aduth replied:
> Moving into `getMultiSelectedBlocks`.

It still needs to be moved?

sgomes replied:
Looks like I had forgotten to push my changes.

sgomes commented:
Thanks for the review, @noisysocks ! I made the changes you suggested and will now explore separating all block keys into their own branch (not just `attributes`), as suggested by @aduth.

aduth commented:
@sgomes Could you put the contents of `git diff | pbcopy` into a Gist so that others could potentially evaluate in the meantime?

sgomes commented:
> @sgomes Could you put the contents of `git diff | pbcopy` into a Gist so that others could potentially evaluate in the meantime?

Here's the WIP diff: https://gist.github.com/sgomes/1b38919a07ba02a0712e390fa6b263cc

sgomes commented:
I spent all of yesterday implementing and validating the proposed changes by @aduth, and a considerable amount of time today trying to understand why they don't perform as expected.

My conclusion so far is that `getBlock` isn't memoizing as expected, but I can't figure out why, as the items that are added to the dependants array stay equal by reference, as far as I can tell.

My suggestion is to review the PR as it is, and leave the proposed changes for later if needed for clarity, which I believe was the main argument. Changes of this nature are very well contained, affecting only the reducer, the selector, and their corresponding tests, so it doesn't seem to me like too much of an issue to change things twice.

sgomes commented:
It appears that `rememo` doesn't handle native-value (e.g. string, boolean) dependants the way I expected it would; that is, by directly comparing them.

It's not clear to me whether this is a bug in `rememo`, but the fact that it tries to record the `lastDependents` leads me to believe it might be. I haven't fully grokked the library code, though, so I could be wrong.

aduth commented:
> It appears that `rememo` doesn't handle native-value (e.g. string, boolean) dependants the way I expected it would; that is, by directly comparing them with their past values, taking into account the dependant function arguments.

Would you be able to put together a minimal example demonstrating the issue? I'm not sure I follow what's unexpected.

FYI, the `rememo` package is my own, so I'd be glad to help clarify or resolve any issues.

There is some distinct treatment (more an optimization) for objects and arrays as dependants, since these can be used as keys of a `WeakMap` to avoid explicitly invalidating a cache value.

sgomes commented:
> Would you be able to put together a minimal example demonstrating the issue? I'm not sure I follow what's unexpected.
> 

No problem, here you go: https://codepen.io/anon/pen/zMmPJo?editors=0011

Let me know if there's anything unclear about the example!

sgomes commented:
I see. That's... problematic. 😅 As far as I can tell that mostly invalidates the state structure you suggested, since it would effectively not be memoizable per-`clientId`, unless I'm missing something?

```
blocks:
  name: { A: 'foo', B: 'bar' },
  attributes: { A: {...}, B: {...} },
  isValid: { A: true, B: true },
  ...
  order: [ ... ]
```

I mean, from what you're saying, memoization in `getBlock` currently only works because all of the dependants in there happen to be non-primitive types, correct?

FWIW, `wp-calypso` has a simple `createSelector` library that achieves memoization based on the dependant function parameters, by generating a cache key based on a simple `join` of the parameters. The restriction there becomes that the parameters should all be primitive types, but that's usually not an issue.

In any case, this all makes me wonder more whether it's possible to keep this PR as-is for the performance benefits, and look into how to achieve the perfect state structure later? 😁

aduth commented:
> I mean, from what you're saying, per-`clientId` memoization in `getBlock` currently only works because all of the dependants in there happen to be non-primitive types, correct?

Yes, that's correct.

> As far as I can tell that mostly invalidates the state structure you suggested, since it would effectively not be memoizable per-`clientId`, unless I'm missing something? At least without custom memoization.

I suppose this is true, yes.

> FWIW, `wp-calypso` has a simple `createSelector` library that achieves memoization based on the dependant function parameters, by generating a cache key based on a simple `join` of the parameters. The restriction there becomes that the parameters should all be primitive types, but that's usually not an issue.

It still destroys the entire cache if dependants references change:

https://github.com/Automattic/wp-calypso/blob/77f67b116e45d1c1cb75e50aae77b01a34fa39c1/client/lib/create-selector/index.js#L102-L104

This is where Calypso's `treeSelect` is meant to serve to have more granular cache invalidation:

https://github.com/Automattic/wp-calypso/tree/master/client/lib/tree-select
https://github.com/Automattic/wp-calypso/pull/20547

`rememo` is effectively implemented as an amalgamation of the two, adopting optimizations similar to `treeSelect` when it detects being possible.

> In any case, this all makes me wonder more whether it's possible to keep this PR as-is for the performance benefits, and look into how to achieve the perfect state structure later? 😁

I'll plan to give this another pass over, but yes, I'd generally be agreeable to a more immediate resolution.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -87,29 +88,52 @@ function mapBlockOrder( blocks, rootClientId = '' ) {
+ * Returns a flattened object with the transformed blocks.
  *
- * @param {Array} blocks Blocks to flatten.
+ * @param {*} blocks Blocks to flatten.
```
Why was the type changed to a wildcard `*` ? It's an `Array`.

sgomes replied:
That's odd, I don't remember changing that. I need to check my IDE settings for JSDoc. Good catch, thanks!

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -87,29 +88,52 @@ function mapBlockOrder( blocks, rootClientId = '' ) {
  *
- * @param {Array} blocks Blocks to flatten.
+ * @param {*} blocks Blocks to flatten.
+ * @param {*} transform Transforming function to be applied to each block.
```
It's not a wildcard `*`, it's a `Function`.

sgomes replied:
👍

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -659,7 +660,9 @@ export const getBlock = createSelector(
 		state.editor.present.blocks.byClientId[ clientId ],
+		state.editor.present.blocks.attributesByClientId[ clientId ],
 		getBlockDependantsCacheBust( state, clientId ),
+		state.editor.present.blocks.order[ clientId ],
```
> Unless I'm mistaken, since `getBlockDependantsCacheBust`calls `getBlock` for each child, we'll recursively depend on the entire set of IDs. Therefore, each `getBlock` only needs to worry about its own direct children, where it comes to `order`.

Isn't this already encompassed by including `getBlockDependantsCacheBust` as a dependant? i.e. this line shouldn't be necesary at all?

sgomes replied:
Correct, it's redundant.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -691,9 +693,16 @@ export const getBlocks = createSelector(
+			];
+		}
+		return [
+			getBlock.getDependants( state, rootClientId ),
```
This is wrong because it will produce a nested array. `getDependants` returns a newly-generated array every time it's called, so the dependants will always be different. This should either (a) spread `...` into the return array or (b) be returned directly as the return value.

sgomes replied:
Excellent catch! Silly mistake; fixed.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,63 +428,114 @@ export const editor = flow( [
-					// Skip update if nothing has been changed. The reference will
-					// match the original block if `reduce` had no changed values.
-					if ( nextAttributes === state[ action.clientId ].attributes ) {
+					// Do nothing if only attributes change.
```
What does / should this do if the update is for `innerBlocks`?

sgomes replied:
Is it valid semantics to modify inner blocks with an `UPDATE_BLOCK` instead of `MOVE_*` and other actions? I assumed not, since that possibility isn't covered in the existing code.

aduth replied:
While I think it's something we should respect, we already don't have consideration for a changed `innerBlocks` from the `order` subtree, so I think it'd be fine to defer.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -361,63 +428,114 @@ export const editor = flow( [
+					};
+
+				case 'REPLACE_BLOCKS':
+					if ( ! action.blocks ) {
```
I know this was inherited, but looking at it, it seems quite wrong. We call `replaceBlocks` with an empty array when, for example, removing a paragraph:

https://github.com/WordPress/gutenberg/blob/b5d4e1527af3f15e350da6be4a57963e51208bec/packages/block-library/src/paragraph/edit.js#L116-L117

If I'm reading this right, this behavior would leave lingering state for the removed block?

sgomes replied:
An empty array is truthy, so it seems to me that this should work correctly, unless I'm missing something?

aduth replied:
> An empty array is truthy, so it seems to me that this should work correctly, unless I'm missing something?

That's a good point. Makes me wonder why the condition exists at all then 🤷‍♂️

sgomes commented:
E2E tests seem to be consistently timing out. Not sure if this reflects a problem in the code or merely CI issues.

sgomes commented:
This time the measured improvement is ~26%.

aduth commented:
There were end-to-end failures across the board last week which have since been resolved. They should be cleared up if you rebase your branch or merge the latest master.

In the meantime, I'll give this another review pass.

sgomes commented:
Squashed all commits to simplify rebasing. Working on that now.

sgomes commented:
Rebased on master and made to work well with `getBlockAttributes`.

iseulde commented:
@sgomes Will this help with garbage collection at random function calls that I'm seeing in master, sometimes multiple times per key press, other times after several key presses?

Sometimes this adds up to over 14MB and I've seen this to take over 130ms.

<img width="440" alt="screen shot 2018-12-04 at 12 33 57" src="https://user-images.githubusercontent.com/4710635/49441084-bb6e8300-f7c5-11e8-8594-57066f1d76d9.png">

<img width="698" alt="screen shot 2018-12-04 at 13 01 10" src="https://user-images.githubusercontent.com/4710635/49441091-c1646400-f7c5-11e8-8805-e8bf61261d52.png">

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -352,63 +419,114 @@ export const editor = flow( [
 						},
 					};
 
+				case 'UPDATE_BLOCK_ATTRIBUTES':
```
Kinda makes me wish we'd normalized `UPDATE_BLOCK_ATTRIBUTES` to use `UPDATE_BLOCK`, but it's also not something I think we ought to be addressing here.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -729,9 +729,14 @@ export const getBlocks = createSelector(
-		state.editor.present.blocks,
-	]
+	( state, rootClientId ) => {
+		if ( ! rootClientId ) {
```
Given the prior noted observation about dependants caching not occurring per arguments set, I'm not sure this is a change we should want to make. Is it even relevant to the changes, or could it at least be considered separately?

sgomes replied:
Nothing like measuring :) I couldn't find any meaningful difference between the two, so I reverted back to the simpler code as per your suggestion.

sgomes committed:
Revert changes to getBlocks dependants

aduth reviewed (approved):
Looks good 👍  Let's hold off on merging until the code freeze is released.

SchneiderSam commented:
Technically I have no idea, but I wonder if it doesn't make sense to add this PR to the 5.0.1 roadmap, especially since [this bug](https://github.com/WordPress/gutenberg/issues/11782#issuecomment-444825896) should be solved with 5.0.1:

![screenshot_9](https://user-images.githubusercontent.com/9355398/49696273-365de200-fba8-11e8-83de-973e219d42ea.png)


If that has nothing to do with it: Sorry ;-)

You are great! Thank you very much for your great work!!!

aduth commented:
Given it is higher risk than other related performance effort, it's been slated for 4.8 (WP 5.0.2) rather than the more-immediate 5.0.1.

youknowriad commented:
Awesome work on this PR @sgomes Thanks.
